File: library/src/main/java/com/pengrad/telegrambot/model/MessageEntity.java
package com.pengrad.telegrambot.model;

import java.io.Serializable;
import java.util.Objects;

/**
 * stas
 * 5/3/16.
 */
public class MessageEntity implements Serializable {
    private final static long serialVersionUID = 0L;

    public enum Type {
        mention, hashtag, cashtag, bot_command, url, email, phone_number, bold, italic, code, pre, text_link,
        text_mention, underline, strikethrough, spoiler, custom_emoji, blockquote, expandable_blockquote
    }

    private Type type;
    private Integer offset;
    private Integer length;
    private String url;
    private User user;
    private String language;
    private String custom_emoji_id;

    private MessageEntity() {
    }

    public MessageEntity(Type type, Integer offset, Integer length) {
        this.type = type;
        this.offset = offset;
        this.length = length;
    }

    public Type type() {
        return type;
    }

    public Integer offset() {
        return offset;
    }

    public Integer length() {
        return length;
    }

    public String url() {
        return url;
    }

    public User user() {
        return user;
    }

    public String language() {
        return language;
    }

    public String customEmojiId() {
        return custom_emoji_id;
    }

    public MessageEntity url(String url) {
        this.url = url;
        return this;
    }

    public MessageEntity user(User user) {
        this.user = user;
        return this;
    }

    public MessageEntity language(String language) {
        this.language = language;
        return this;
    }

    public MessageEntity customEmojiId(String custom_emoji_id) {
        this.custom_emoji_id = custom_emoji_id;
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MessageEntity that = (MessageEntity) o;
        return type == that.type &&
                Objects.equals(offset, that.offset) &&
                Objects.equals(length, that.length) &&
                Objects.equals(url, that.url) &&
                Objects.equals(user, that.user) &&
                Objects.equals(language, that.language) &&
                Objects.equals(custom_emoji_id, that.custom_emoji_id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, offset, length, url, user, language, custom_emoji_id);
    }

    @Override
    public String toString() {
        return "MessageEntity{" +
                "type=" + type +
                ", offset=" + offset +
                ", length=" + length +
                ", url='" + url + '\'' +
                ", user=" + user +
                ", language='" + language + '\'' +
                ", custom_emoji_id='" + custom_emoji_id + '\'' +
                '}';
    }
}

File: core/src/commonMain/kotlin/com/pengrad/telegrambot/model/MessageEntity.kt
package com.pengrad.telegrambot.model

import java.io.Serializable

/**
 * stas
 * 5/3/16.
 */
class MessageEntity : Serializable {

    private val type: Type? = null
    private val offset: Int? = null
    private val length: Int? = null
    private val url: String? = null
    private val user: User? = null

    enum class Type {
        mention, hashtag, cashtag, bot_command, url, email, phone_number, bold, italic, code, pre, text_link,
        text_mention
    }

    fun type(): Type? {
        return type
    }

    fun offset(): Int? {
        return offset
    }

    fun length(): Int? {
        return length
    }

    fun url(): String? {
        return url
    }

    fun user(): User? {
        return user
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false

        val that = o as MessageEntity?

        if (type != that!!.type) return false
        if (if (offset != null) offset != that.offset else that.offset != null) return false
        if (if (length != null) length != that.length else that.length != null) return false
        if (if (url != null) url != that.url else that.url != null) return false
        return if (user != null) user == that.user else that.user == null

    }

    override fun hashCode(): Int {
        var result = type?.hashCode() ?: 0
        result = 31 * result + (offset?.hashCode() ?: 0)
        result = 31 * result + (length?.hashCode() ?: 0)
        result = 31 * result + (url?.hashCode() ?: 0)
        result = 31 * result + (user?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "MessageEntity{" +
                "type=" + type +
                ", offset=" + offset +
                ", length=" + length +
                ", url='" + url + '\''.toString() +
                ", user=" + user +
                '}'.toString()
    }

    companion object {
        private const val serialVersionUID = 0L
    }
}
